### PR TITLE
M3 lz 34

### PR DESCRIPTION
**Pull request by Eternalynx**
Added highlights for mockup surveys. To create them, click Add Highlight button under the question, then click move mouse click. To save click the Save button (bottom right corner), to cancel click Cancel. To see highlights click the Toggle Highlights button in MockupSurvey view. Click again to hide.

### CODE REVIEW

**Comment by agsta** [comment]
@Eternalynx can you switch the master branch to develop?

**Comment by agsta** [comment]
- Currently, the highlights are saved with `mockupId `+ `questionId`, but not with `surveyId `- as a result in a survey view all highlights created for mockup are visible.
- The 'Save' / 'Cancel' buttons could be somewhere higher, they are not always visible for smaller screens
- Maybe we could allow user to set the height/width while uploading the mockup? This way the iframe could be set to fit the content and we wouldn't have the scrollbar issue (however this wouldn't solve the problem with changing pages in the prototype). But it's just a suggestion, we probably don't have time for that

**Comment by mimagiera** [comment]
> * Currently, the highlights are saved with `mockupId `+ `questionId`, but not with `surveyId `- as a result in a survey view all highlights created for mockup are visible.
> * The 'Save' / 'Cancel' buttons could be somewhere higher, they are not always visible for smaller screens
> * Maybe we could allow user to set the height/width while uploading the mockup? This way the iframe could be set to fit the content and we wouldn't have the scrollbar issue (however this wouldn't solve the problem with changing pages in the prototype). But it's just a suggestion, we probably don't have time for that

2.  'Save' / 'Cancel' buttons are overlapping with other content, maybe placi it next yo `Add highlight` button
image
3. We could also always set height of iframe to "max" of mockup and probably never have problem with scrolling it. But as you said, another problems will sill exist and we don't have time

**Comment by mimagiera** [comment]
Another thing, this text and button should be alligned vertically (or maybe better this text should be a tooltip as @agsta suggests)
image

File: front/src/components/AddHighlightPopup.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        super(props);
+        this.state = {
+            serverUrl: process.env.REACT_APP_SERVER_URL,
+            index: this.props.index,
+            mockupId: this.props.mockupId,
+            showPopup: false,
+            isDrawing: false,
+            posX: 0,
+            posY: 0,
+            hght: 0,
```
Comment (mimagiera) [inline front/src/components/AddHighlightPopup.js]: why not `height`?

File: front/src/components/AddMockupSurvey.js [hunk 2]
```diff
@@ -306,6 +328,13 @@ class AddMockupSurvey extends Component {
                                                                 >
                                                                 </Input>
                                                             </Col>
+                                                            <Col>
+                                                                Highlight doesn't work with scrolling mockups
```
Comment (mimagiera) [inline front/src/components/AddMockupSurvey.js]: this text could be a const, don't repeat it 4 times

**Comment by agsta** [comment]
> Another thing, this text and button should be alligned vertically (or maybe better this text should be a tooltip as @agsta suggests)
> image

See: https://reactstrap.github.io/components/tooltips/

File: front/src/components/AddMockupSurvey.js [hunk 3]
```diff
@@ -339,6 +368,13 @@ class AddMockupSurvey extends Component {
                                                                     className="btn btn-primary btn-margin-top">Add Statement
                                                                 </Button>
                                                             </Col>
+                                                            <Col>
```
Comment (mimagiera) [inline front/src/components/AddMockupSurvey.js]: actually creating this column could be extracted to a function. e.g. (maybe with better function name):
```
    getHighlightCol(index) {
        return <Col>
            Highlight doesn't work with scrolling mockups
            <Button
                onClick={(event) => this.addHighlight(event, index)}
                className="btn btn-primary btn-margin-top">Add Highlight
            </Button>
        </Col>;
    }
```

File: front/src/components/HighlightPopup.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import React, {Component} from "react";
+
+import AuthenticationService from "../service/AuthenticationService";
+import axios from "axios";
+import {Layer, Stage, Rect, Group, Text} from "react-konva";
+
+
+
```
Comment (mimagiera) [inline front/src/components/HighlightPopup.js]: sooo many empty lines

File: front/src/components/HighlightPopup.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        super(props);
+        this.state = {
+            serverUrl: process.env.REACT_APP_SERVER_URL,
+            mockupId: this.props.mockupId,
+            showPopup: false,
+            highlights: []
+        };
+
+
```
Comment (mimagiera) [inline front/src/components/HighlightPopup.js]: sooo many empty lines

File: src/main/java/lz/modelTesting/controllers/HighlightsController.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
+package lz.modelTesting.controllers;
+
+
```
Comment (mimagiera) [inline src/main/java/lz/modelTesting/controllers/HighlightsController.java]: sooo many empty lines

File: src/main/java/lz/modelTesting/controllers/HighlightsController.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import javax.servlet.http.HttpServletRequest;
+import java.util.List;
+
+@RestController
+@RequestMapping("/highlights")
+@CrossOrigin(origins = "*", allowedHeaders = "*")
+public class HighlightsController {
+
+    private final static String MOCKUP_ID = "mockupId";
```
Comment (mimagiera) [inline src/main/java/lz/modelTesting/controllers/HighlightsController.java:17]: that could be reused from `SurveysController` and even better extracted to another place (another class) and be used here and in `SurveysController` from that another place

File: src/main/java/lz/modelTesting/controllers/HighlightsController.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    private Highlight createHighlightFromRequest(HttpServletRequest request) {
+        String mockupId = request.getParameter(MOCKUP_ID);
+        String questionNumber = request.getParameter(QUESTION_NUMBER);
+        String width = request.getParameter(WIDTH);
+        String height = request.getParameter(HEIGHT);
+        String posX = request.getParameter(POS_X);
+        String posY = request.getParameter(POS_Y);
+
+
```
Comment (mimagiera) [inline src/main/java/lz/modelTesting/controllers/HighlightsController.java]: sooo many empty lines

File: src/main/java/lz/modelTesting/documents/Highlight.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    private String width;
+    @NonNull
+    private String height ;
+    @NonNull
+    private String posX;
+    @NonNull
+    private String posY;
+
+
+
```
Comment (mimagiera) [inline src/main/java/lz/modelTesting/documents/Highlight.java]: sooo many empty lines

File: src/main/java/lz/modelTesting/controllers/HighlightsController.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.List;
+
+@RestController
+@RequestMapping("/highlights")
+@CrossOrigin(origins = "*", allowedHeaders = "*")
+public class HighlightsController {
```
Comment (mimagiera) [inline src/main/java/lz/modelTesting/controllers/HighlightsController.java:15]: as @agsta said, highlights should be also taken by `surveyId`

**Review by mimagiera — CHANGES_REQUESTED** [review]
* pls format code properly (ctrl alt L) and imports (ctrl alt O)
* delete all unnecessary empty lines
* change highlight entity to also contain surveyId
* make another name refactors / extract redundant code to functions

**Comment by Eternalynx** [comment]
Problem with changing mockupId to surveyId - at the point of creating highlight surveyId doesn't exist because survey isn't saved yet. Should I try to manually create surveyId before? But how would I do that?

**Comment by Eternalynx** [comment]
Also, can I place save and cancel buttons on the bar next to fullscreen and show preview? I cannot find any other place that wouldn't interfere with anything

**Comment by mimagiera** [comment]
> Problem with changing mockupId to surveyId - at the point of creating highlight surveyId doesn't exist because survey isn't saved yet. Should I try to manually create surveyId before? But how would I do that?

How about setting surveyId in highlight during saving whole survey to database?

**Comment by mimagiera** [comment]
> Also, can I place save and cancel buttons on the bar next to fullscreen and show preview? I cannot find any other place that wouldn't interfere with anything

MAybe next to button "Add highlight"?

**Comment by Eternalynx** [comment]
There are multiple add highlight buttons with different positions depending on question type

**Comment by mimagiera** [comment]
> There are multiple add highlight buttons with different positions depending on question type

maybe dynamically create those buttons next to this "add highlight" which was clicked

**Comment by Eternalynx** [comment]
> > There are multiple add highlight buttons with different positions depending on question type
> 
> maybe dynamically create those buttons next to this "add highlight" which was clicked

I have no idea how to get position of clicked button



> > Problem with changing mockupId to surveyId - at the point of creating highlight surveyId doesn't exist because survey isn't saved yet. Should I try to manually create surveyId before? But how would I do that?
> 
> How about setting surveyId in highlight during saving whole survey to database?

Should I pass to the backend list of highlight ids then? How about procreating a survey in componentDidMount and updating it on submit? Would that work?

**Review by agsta — APPROVED** [review]

**Review by mimagiera — APPROVED** [review]

**Pull request merged**